pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/DataBlockHeaderImpl.java

```diff
@@ -0,0 +1,124 @@
+    // It is a sequence of bytes used to identify the start of a block.
+    private static final int dataBlockMagicWord = 0xDBDBDBDB;
+    // This is bigger than header size. Leaving some place for alignment and future enhancement.
+    // Payload use this as the start offset.
+    private static final int dataBlockHeaderAlign = 128;
+    private static final byte padding = 0;
+    // The size of this header.
+    private static final int dataBlockHeaderSize = 4 /* magic word */
+        + 4 /* index block length */
+        + 8 /* first entry id */;
+
+    public static DataBlockHeaderImpl of(int blockLength, long firstEntryId) {
+        return new DataBlockHeaderImpl(blockLength, firstEntryId);
+    }
+
+    // Construct DataBlockHeader from InputStream, which contains `dataBlockHeaderAlign` bytes readable.
+    public static DataBlockHeader fromStream(InputStream stream) throws IOException {
+        DataInputStream dis = new DataInputStream(stream);
+        int magic = dis.readInt();
+        if (magic != dataBlockMagicWord) {
+            throw new IOException("Data block header magic word not match.");
+        }
+
+        int blockLen = dis.readInt();
+        long firstEntryId = dis.readLong();
+
+        // verify padding
+        for (int index = 0; index < dataBlockHeaderAlign - dataBlockHeaderSize; index ++) {
+            if (dis.readByte() != padding) {
+                throw new IOException("Data block header magic word not match.");
+            }
+        };
+
+        return new DataBlockHeaderImpl(blockLen, firstEntryId);
+    }
+
+    private final int blockLength;
+    private final long firstEntryId;
+
+    static public int getBlockMagicWord() {
+        return dataBlockMagicWord;
+    }
+
+    static public int getDataStartOffset() {
+        return dataBlockHeaderAlign;
+    }
+
+    @Override
+    public int getBlockLength() {
+        return this.blockLength;
+    }
+
+    @Override
+    public long getFirstEntryId() {
+        return this.firstEntryId;
+    }
+
+    @Override
+    public int getHeaderSize() {
+        return dataBlockHeaderAlign;
+    }
+
+    public DataBlockHeaderImpl(int blockLength, long firstEntryId) {
+        this.blockLength = blockLength;
+        this.firstEntryId = firstEntryId;
+    }
+
+    /**
+     * Get the content of the data block header as InputStream.
+     * Read out in format:
+     *   [ magic_word -- int ][ block_len -- int ][ first_entry_id  -- long] [padding zeros]
+     */
+    @Override
+    public InputStream toStream() throws IOException {
+        int headerSize = 4 /* magic word */
+            + 4 /* index block length */
+            + 8 /* first entry id */;
+
+        ByteBuf out = PooledByteBufAllocator.DEFAULT.buffer(dataBlockHeaderAlign, dataBlockHeaderAlign);
+        out.writeInt(dataBlockMagicWord)
+            .writeInt(blockLength)
+            .writeLong(firstEntryId)
+            .writeBytes(new byte[dataBlockHeaderAlign - headerSize]);
```

Comment:
the padding will always be the same size, so define PADDING = new byte[x] as a constant in the class. When reading back, you can skip PADDING.length also.